amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/contexts/WebApiContext.scala

```diff
@@ -50,9 +55,21 @@ abstract class WebApiContext(val loc: String,
 
   var jsonSchemaRefGuide: JsonSchemaRefGuide = JsonSchemaRefGuide(loc, refs)(this)
 
+  var indexCache: mutable.Map[String, AstIndex] = mutable.Map[String, AstIndex]()
+
   def setJsonSchemaAST(value: YNode): Unit = {
+    val location = value.sourceName
```

Comment:
For me the best option is that the Fragment holds it's index. This has several problems though:
- The cache and index would both have to have concurrency control.
- A JSON File can be interpreted using several schemas.

Although this improves performance by saving indexing several times when document A parses a type in Document B we still have unnecesary processing.